### ISSUE TITLE
Fail to build with LTO enabled

### ISSUE THREAD

**Issue by r-value** [gnuradio/gnuradio#5635]
### What happened?

Making package for Arch Linux, with LTO enabled. `ld` complains that:

```
/usr/bin/ld: libgnuradio-digital.so.3.9.4.0: undefined reference to `gr::digital::map_bb_impl::s_map_size'
```

### System Information

OS: Arch Linux
GR Installation Method: build from source

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/11.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++,d --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror --with-build-config=bootstrap-lto --enable-link-serialization=1 gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (GCC)
$ ld -v
GNU ld (GNU Binutils) 2.38
```

### GNU Radio Version

3.9  (maint-3.9)

### Specific Version

3.9.4.0

### Steps to Reproduce the Problem

Build script:

```
  export PYTHON=python3
  cd "gnuradio-3.9.4.0"
  cmake -DCMAKE_INSTALL_PREFIX=/usr \
    -DPYTHON_EXECUTABLE=$(which python3) \
    -DPYTHON_INCLUDE_DIR=/usr/include/python3.10 \
    -DPYTHON_LIBRARY=/usr/lib/libpython3.10.so \
    -DGR_PYTHON_DIR=/usr/lib/python3.10/site-packages \
    -DENABLE_INTERNAL_VOLK=OFF \
    -DENABLE_GRC=ON \
    -DENABLE_GR_QTGUI=ON \
    -DQWT_LIBRARIES=/usr/lib/libqwt.so \
    -Wno-dev \
    -B build \
    -S .
  make -C build
```

### Relevant log output

https://gist.github.com/r-value/5b6fc6e54cdd92b34fce4128ebb3b8ff

**Comment by rear1019** [comment]
https://github.com/gnuradio/gnuradio/blob/7c989ae932357801d1eadbf68eef1fdb3b3fbfd8/gr-digital/lib/map_bb_impl.h#L23

This declaration alone (without further code) only works since C++17; GNU Radio 3.9.4.0 is compiled using C++14. I wonder why it works at all, but this might be just another subtlety in C++… Anyway, the following patch should fix the issue for older C++ revisions (not tested).

```diff
diff --git a/gr-digital/lib/map_bb_impl.cc b/gr-digital/lib/map_bb_impl.cc
index d1f5b134f..5270bd891 100644
--- a/gr-digital/lib/map_bb_impl.cc
+++ b/gr-digital/lib/map_bb_impl.cc
@@ -18,6 +18,8 @@
 namespace gr {
 namespace digital {
 
+constexpr size_t map_bb_impl::s_map_size;
+
 map_bb::sptr map_bb::make(const std::vector<int>& map)
 {
     return gnuradio::make_block_sptr<map_bb_impl>(map);
```

**Comment by marcusmueller** [comment]
Calling @ThomasHabets on this one!

**Comment by ThomasHabets** [comment]
Yeah, @rear1019 is right. This is required in C++14 "if it is odr-used" (12.2.3.2 (class.static.data) paragraph 3), and actually deprecated as unnecessary in C++17 (D1 (depr.static_constexpr)).

I don't quite understand why LTO makes it ODR-used. My reading of 6.2 (basic.def.odr) paragraph 3 is that this should fall under the exception as it seems to me that the lvalue-to-rvalue conversion should yield a constant expression. I mean it *is* a constant expression, and we're using it just as-is.

In any case adding the definition, as @rear1019 said, can't hurt, and should help. Because from the logs it sure looks like the linker expects it.

**Comment by r-value** [comment]
Seems that the patch from @rear1019 works. I'd appreciate it if this patch could be merged and released.

I'd like to leave this issue open until the changes are released.

**Comment by marcusmueller** [comment]
solved by std upgrade in 3.10
